README.md
# commitlint-config-non-conventional

> Shared [commitlint] config to enforce a good commit message

This config disllows the usage of [conventinal style commits][conventionalcommits]. Sometimes all you need is just a proper commit messages with rules enforced.

[![Tests](https://github.com/sibiraj-s/commitlint-config-non-conventional/actions/workflows/tests.yml/badge.svg)](https://github.com/sibiraj-s/commitlint-config-non-conventional/actions/workflows/tests.yml)
[![Version](https://badgen.net/npm/v/commitlint-config-non-conventional)](https://npmjs.com/package/commitlint-config-non-conventional)
[![License](https://badgen.net/npm/license/commitlint-config-non-conventional)](https://github.com/sibiraj-s/commitlint-config-non-conventional/blob/master/LICENSE)
![Node](https://badgen.net/npm/node/commitlint-config-non-conventional)

## Rules

A commit message or commit title should

- not be empty
- be sentence cased
- contain atleast 10 characters
- not exceed 50 characters
- not end with full stop

If the commit message has a body and footer. They should have 2 blank leading lines and should not exceed 72 characters. The description(body) should be sentence cased.

## Installation

```bash
npm i -D @commitlint/cli commitlint-config-non-conventional
```

## Usage

```js
// commitlint.config.js

module.exports = {
  extends: 'non-conventional',
};
```

### Configuring with husky

```bash
# Install husky
npm i -D husky

# Activate husky hooks
npx husky install

# Add commit-msg hook
npx husky add .husky/commit-msg 'npx --no -- commitlint --edit "$1"'
```

Checkout the [husky documentation][install-husky] on how to automatically install git hooks post installation.

[commitlint]: https://commitlint.js.org
[conventionalcommits]: https://www.conventionalcommits.org/
[install-husky]: https://typicode.github.io/husky/#/?id=manual

### Useful links

- [How to Write a Git Commit Message](https://chris.beams.io/posts/git-commit/)


CHANGELOG.md
# CHANGELOG

All notable changes to this project will be documented in this file.

## v1.0.1 (2021-12-08)

- add rules to README ([5f021ee](https://github.com/sibiraj-s/commitlint-config-non-conventional/commit/5f021ee))

## v1.0.0 (2021-10-31)

- Initial release. Shareable commitlint config.


test.js
const { test, expect } = require('@jest/globals');
const lint = require('@commitlint/lint').default;
const { rules } = require('.');

const tests = [
  {
    name: 'a valid oneline commit message',
    message: 'Great commit summaries contain fewer than 50 chars',
    valid: true,
    errors: [],
  },
  {
    name: 'a detailed commit',
    message: 'Great commit summaries contain fewer than 50 chars'
      + '\n\n'
      + 'More detailed explanatory text, if necessary. Wrap it to about 72\n'
      + 'characters or so. In some contexts, the first line is treated as the\n'
      + 'subject of the commit and the rest of the text as the body\n'
      + '\n\n'
      + 'Fixes #1',
    valid: true,
    errors: [],
  },
  {
    name: 'message should be sentence cased',
    message: 'fix error with something',
    valid: false,
    errors: ['header-case'],
  },
  {
    name: 'message should not exceed 50 chars',
    message: 'This is a long commit message which is not accepted by given rules',
    valid: false,
    errors: ['header-max-length'],
  },
  {
    name: 'message should contain at least 10 chars',
    message: 'Invalid',
    valid: false,
    errors: ['header-min-length'],
  },
  {
    name: 'should not allow conventional commit type',
    message: 'fix: error with something',
    valid: false,
    errors: ['type-empty', 'subject-empty', 'header-case'],
  },
  {
    name: 'should not allow conventional commit with scope',
    message: 'fix(scope): error message with type and scope',
    valid: false,
    errors: ['header-case', 'type-empty', 'scope-empty', 'subject-empty'],
  },
  {
    name: 'body should have a leading blank line',
    message: `Great commit summaries contain fewer than 50 chars
    Body should have leading blank line to seperate from the subject
    `,
    valid: false,
    errors: ['body-leading-blank'],
  },
  {
    name: 'body should have a leading blank line',
    message: 'Great commit summaries contain fewer than 50 chars'
      + '\n\n'
      + 'A detailed explanatory decscription on the change, should be wrapped within 72 characters',
    valid: false,
    errors: ['body-max-line-length'],
  },
  {
    name: 'footer should have a leading blankline',
    message: 'Great commit summaries contain fewer than 50 chars'
      + '\n\n'
      + 'A detailed explanatory decscription on the change'
      + '\n'
      + 'Fixes #1',
    valid: false,
    errors: ['footer-leading-blank'],
  },
  {
    name: 'footer should have a leading blankline',
    message: 'Great commit summaries contain fewer than 50 chars'
      + '\n\n'
      + 'A detailed explanatory decscription on the change'
      + '\n\n'
      + 'BREAKING CHANGE: Footer comesFooter sentence should not exceed 72 characters',
    valid: false,
    errors: ['footer-max-line-length'],
  },
];

test.each(tests)('$name', async ({ message, ...t }) => {
  const result = await lint(message, rules);

  expect(result.valid).toBe(t.valid);
  expect(t.errors.length).toEqual(result.errors.length);

  const resultErrors = result.errors.map((err) => err.name);
  expect(t.errors).toEqual(expect.arrayContaining(resultErrors));
});


index.js
// https://commitlint.js.org/#/reference-rules

module.exports = {
  rules: {
    // disable type
    'type-empty': [2, 'always'],

    // disable scope
    'scope-empty': [2, 'always'],

    // diable subject
    // since type and scope are disabled, first line of the header is the subject
    'subject-empty': [2, 'always'],

    // header
    'header-case': [2, 'always', 'sentence-case'],
    'header-min-length': [2, 'always', 10],
    'header-max-length': [2, 'always', 50],
    'header-full-stop': [2, 'never'],

    // body
    'body-leading-blank': [2, 'always'],
    'body-case': [2, 'always', 'sentence-case'],
    'body-max-line-length': [2, 'always', 72],

    // footer
    'footer-leading-blank': [2, 'always'],
    'footer-max-line-length': [2, 'always', 72],
  },
  helpUrl: '',
};
